File: dotnet/M3Mobile.M3Sdk.Xamarin/Startup/ProjectMediaRequest.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using Android.Content;
using Android.OS;
using M3Sdk.Xamarin.Internal;

namespace M3Sdk.Xamarin.Startup
{
    internal sealed class ProjectMediaRequest
    {
        private const int TimeoutMillis = 3000;
        private readonly Context _context;
        private readonly string _packageName;

        internal ProjectMediaRequest(Context context, string packageName)
        {
            _context = context;
            _packageName = packageName;
        }

        internal async Task<ProjectMediaResult> ResultAsync(CancellationToken cancellationToken)
        {
            var resultSource = new TaskCompletionSource<ProjectMediaResult>();
            var handler = new IncomingHandler(Looper.MainLooper, message =>
            {
                var errorMessage = message.Data == null
                    ? string.Empty
                    : message.Data.GetString(Constants.StartUp.ExtraProjectMediaErrorMessage);
                resultSource.TrySetResult(ProjectMediaContract.Result(message.What, errorMessage));
            });
            var messenger = new Messenger(handler);
            var timeoutSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
            timeoutSource.CancelAfter(TimeoutMillis);
            var registration = timeoutSource.Token.Register(() =>
            {
                if (cancellationToken.IsCancellationRequested)
                {
                    resultSource.TrySetCanceled();
                    return;
                }

                resultSource.TrySetException(
                    new TimeoutException("Timed out waiting for the StartUp PROJECT_MEDIA response."));
            });

            try
            {
                var intent = new Intent(Constants.StartUp.RequestSystem)
                    .SetPackage(Constants.StartUp.PackageName)
                    .PutExtra(Constants.StartUp.TypeSetting, Constants.StartUp.TypeProjectMedia)
                    .PutExtra(Constants.StartUp.ExtraProjectMediaPackage, _packageName)
                    .PutExtra(Constants.StartUp.ExtraProjectMediaMessenger, messenger);
                _context.SendBroadcast(intent);
                return await resultSource.Task.ConfigureAwait(false);
            }
            finally
            {
                registration.Dispose();
                timeoutSource.Dispose();
                messenger.Dispose();
                handler.Dispose();
            }
        }

        private sealed class IncomingHandler : Handler
        {
            private readonly Action<Message> _message;

            internal IncomingHandler(Looper looper, Action<Message> message)
                : base(looper)
            {
                _message = message;
            }

            public override void HandleMessage(Message message)
            {
                _message(message);
            }
        }
    }
}

File: feature/startup/src/main/java/net/m3mobile/feature/startup/requester/permission/ProjectMediaRequest.kt
package net.m3mobile.feature.startup.requester.permission

import android.content.Context
import android.content.Intent
import android.os.Handler
import android.os.Looper
import android.os.Messenger
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withTimeout
import net.m3mobile.feature.startup.constants.ExtraKey
import net.m3mobile.feature.startup.constants.RequestAction
import net.m3mobile.feature.startup.constants.TypeKey
import net.m3mobile.feature.startup.constants.TypeValue
import net.m3mobile.feature.startup.params.ProjectMediaResult
import net.m3mobile.feature.startup.params.projectMediaResult
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

internal class ProjectMediaRequest(
    private val context: Context,
    private val packageName: String,
) {
    suspend fun result(): ProjectMediaResult =
        withTimeout(TIMEOUT_MILLIS) {
            suspendCancellableCoroutine { continuation ->
                val messenger = Messenger(
                    Handler(Looper.getMainLooper()) { message ->
                        if (!continuation.isActive) return@Handler true

                        continuation.resume(
                            projectMediaResult(
                                message.what,
                                message.data.getString(ExtraKey.PROJECT_MEDIA_ERROR_MESSAGE),
                            ),
                        )
                        true
                    },
                )

                val intent = Intent(RequestAction.SYSTEM)
                    .setPackage(STARTUP_PACKAGE_NAME)
                    .putExtra(TypeKey.SETTING, TypeValue.PROJECT_MEDIA)
                    .putExtra(ExtraKey.PROJECT_MEDIA_PACKAGE, packageName)
                    .putExtra(ExtraKey.PROJECT_MEDIA_MESSENGER, messenger)

                try {
                    context.sendBroadcast(intent)
                } catch (e: Exception) {
                    if (continuation.isActive) continuation.resumeWithException(e)
                }
            }
        }

    private companion object {
        const val STARTUP_PACKAGE_NAME = "com.m3.startup"
        const val TIMEOUT_MILLIS = 3_000L
    }
}
